File: Linked Lists/234. Palindrome Linked List/234. Palindrome Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function (head) {
  // Find middle
  let fast = head;
  let slow = head;
  while (fast && fast.next) {
    fast = fast.next.next;
    slow = slow.next;
  }

  // Reverse second half
  let prev = null;
  while (slow) {
    const temp = slow.next;
    slow.next = prev;
    prev = slow;
    slow = temp;
  }

  // Compare the two halves
  let l = head;
  let r = prev;
  while (r) {
    if (l.val !== r.val) return false;
    l = l.next;
    r = r.next;
  }

  return true;
};

// Time: O(n)
// Space: O(1)

File: Linked Lists/234. Palindrome Linked List/234. Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        # Find middle
        fast = slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        
        # Reverse second half
        prev = None
        while slow:
            temp = slow.next
            slow.next = prev
            prev, slow = slow, temp
        
        # Check compare first and second halves
        l, r = head, prev
        while r:
            if l.val != r.val:
                return False
            l, r = l.next, r.next
        
        return True

# Time: O(n)
# Space: O(1)